mozjs/js/rust/magic_codegen/src/lib.rs

```diff
@@ -37,62 +35,70 @@ fn match_struct(ast: &syn::DeriveInput, variant: &syn::VariantData) -> quote::To
         .split('_')
         .next()
         .expect("Should have a '_' in the magic dom struct name");
+    let from_obj_err = quote::Ident::from(format!("b\"Fail to construct {} from JS object\\0\" as *const u8 as *const libc::c_char", name));
+    let constructor = quote::Ident::from(format!("{}_constructor", name));
     let name_field = quote::Ident::from(format!("b\"{}\\0\"", name));
-    let js_class = quote::Ident::from(format!("{}_CLASS", name.to_uppercase()));
+    let name_upper = name.to_uppercase();
+    let js_class = quote::Ident::from(format!("{}_CLASS", name_upper));
+    let ps_arr = quote::Ident::from(format!("{}_PS_ARR", name_upper));
     let name = quote::Ident::from(name);
     let MagicStructCode {
         getters,
         setters,
         js_getters,
         js_setters,
-        js_prop_spec
+        js_prop_spec,
+        get_callargs,
+        setter_invocations,
     } = get_magic_struct_code(&name, variant);
     let spec_size = js_prop_spec.len() + 1;
     let test_fn_name = quote::Ident::from(format!("test_{}_magic_layout()", name));
     let num_reserved_slots = quote::Ident::from(format!("{}", js_prop_spec.len()));
+    let arg_num_err = quote::Ident::from(format!("b\"constructor requires exactly {} arguments\\0\".as_ptr() as *const libc::c_char", js_prop_spec.len()));
 
     quote! {
         extern crate libc;
-        use js::jsapi;
-        use js::jsapi::root::*;
-        use js::rust::{GCMethods, RootKind};
-        use js::glue::CreateCallArgsFromVp;
-        use js::jsval::DoubleValue;
-        use js::rust::ToNumber;
-        use js::conversions::{ConversionResult, FromJSValConvertible, ToJSValConvertible};
-        use js::jsval;
+        use self::super::*;
+        use self::super::super::jsapi;
+        use self::super::super::jsapi::root::*;
+        use self::super::super::rust::{GCMethods, RootKind, maybe_wrap_value};
+        use self::super::super::glue::CreateCallArgsFromVp;
+        use self::super::super::jsval::{DoubleValue, ObjectOrNullValue, ObjectValue};
+        use self::super::super::rust::ToNumber;
+        use self::super::super::conversions::{ConversionResult, FromJSValConvertible, ToJSValConvertible};
+        use self::super::super::jsval;
```

Comment:
We should keep using absolute imports since we can't control where the macro is invoked, and therefore don't know the correct relative paths to use.